### ISSUE TITLE
Properly managing network disconnection while in a call to reset UI and streams

### ISSUE THREAD

**Issue by desnoexa** [priologic/easyrtc#279]
When testing media disconnection on outgoing calls between 2 peers, we face the following issue.
Environment : Firefox / Chrome or Chrome / Chrome, no matter who is the caller/callee.

Call is established between A and B, A is the caller.
A lost its connection. B will detect it by the ICE error.
The problem is that on B side, nothing can be updated on the UI and the local media streams are still up, because no failure feedback is raised (in the contrary of the caller side with the oniceconnectionstatechange "failed" state that will raise the failureCb callback).

On B side, the callback triggered in this disconnection case is the one set in function doAnswerBody in the errorCallBack of the buildPeerConnection.
It only triggers a log without allowing user to have any feedback and disabling him from closing its local media.
A partial fix could be to add the following code after the log (self.showError(self.errCodes.SYSTEM_ERR, message) of the buildPeerConnection):
// Hangup call if error on media, maybe for other errors too?
if (message === self.errCodes.NOVIABLEICE) {
  closePeer(caller);
  for (i = 0; i < streamNames.length; i++) {
    closeLocalMediaStreamByName(streamNames[i]);
  }
}
What do you think of that ?

Another issue following the first fix is that, depending on which browser the code is running (issue seen on Firefox), closePeer function may fails with an exception (‘RTCPeerConnection is gone (did you enter Offline mode?’).
To allow the user to have a specific feedback, the following code could also be added to closePeer function in the catch section :

} catch (err) {
    logDebug("peer " + otherUser + " close failed:" + err);
    if (self.onPeerFailing) {
        self.onPeerFailing(otherUser)
    }
} finally {

Hope this helps to figure out what’s going on…

Thanks

Xavier

**Comment by EricDavies** [comment]
Xaxier, can you describe the sequence of steps you need to reproduce this problem with one of the supplied demo programs?

**Comment by desnoexa** [comment]
I made a test on https://demo.easyrtc.com/demos/demo_audio_only.html to explain exactly the problem / lack.
Consider A and B are connected on the demo page. A calls B et B accepts.
B is disconnected from the network.
1) After a while, I will enter in the failureCb of the buildPeerConnection of the doAnswerBody method on B's side. The problem is that I only have the showError today in easyrtc.js. So I can't decide to manage properly my UI or my local streams (if I want) because I am not aware of this problem in my application.
==> We think that after the showError in this case, we should add a call to an applicative callback (the errorHandler of the acceptCall ??) to be able to clean the UI if necessary.

In fact, in the same configuration, if B calls A and B is disconnected, the behavior is slightly different, because in easyrtc.js we have the call to the failureCb in the iceconnectionstatechange so we can manage the UI modification through this callback.

2) On A's side, in the demo, nothing happens on the UI. We suppose that in fact you don't have any peerFailingListener implemented, that's why nothing is updated on the UI with the disconnection.
We just have the list of connected users updated through the RoomOccupantListener...

For the second point, it is pointed in Firefox specially when we call closePeer method.
In fact, in FF, we wannot make the difference between a problem on the peer due to simple disconnection or to a normal call hangup.
Adding the self.onPeerFailing in the catch enables to manage specifically the problem of disconnection instead of going immediately to the onPeerClosed of the finally.

**Comment by hthetiot** [comment]
related #272

**Comment by EricDavies** [comment]
I'll add this in sometime this week.

**Comment by EricDavies** [comment]
Xavier: closing the local media streams on a peer connection error would be a bad idea because that media stream wasn't necessarily opened for the peer connection. The most obvious case being: if you've got a multiparty call going, and one peer connection fails, you don't want it to shut down the media stream to the other peer connections.

I've got a modified version of easyrtc.js in which I've added an optional third argument to the helper function passed to the application supplied acceptChecker, that argument being a callFailureCB. That will give us symmetry with the Calling side. Once I verify it hasn't broken anything (I'm seeing some weird behaviour with demo_audio_video) , I'll put a copy up for you to try before we merge it into the code base.

**Comment by EricDavies** [comment]
Xavier: here is the replacement api/easyrtc.js file for you to try: https://gist.github.com/EricDavies/edb23f8ed5ad12ccda713812119caa00 .
It's been tested to the extent that it doesn't cause problems under normal conditions.
Just to be clear: you'll need to modify your acceptChecker functions to get any advantage from it.

easyrtc.setAcceptChecker(function(easyrtcid, responsefn) {
        function myCallFailureCB(errorCode, errorText) {
               console.log("call failed ", easyrtcid, errorcode, errorText);
        }
        responsefn(true, myOutgoingMediaStreams, myCallFailureCB);
        document.getElementById("connectbutton_" + easyrtcid).style.visibility = "hidden";
    });

**Comment by desnoexa** [comment]
I see no difference in the URL you sent with what is currently in the beta branch :( Maybe I missed something...

**Comment by EricDavies** [comment]
Nope, I missed something, like the correct directory. Sigh, life happens.
Here is a new gist.   https://gist.github.com/EricDavies/5f5ab019de7f750fcd57f8adf87061a6
As a quick confirmation, if you look at line 7819, you'll see an extra parameter for the call failure.
There is some other stuff in there I've been trying out (iceRestart), but it shouldn't interfere.
